polar-core/src/partial/constraints.rs

```diff
@@ -0,0 +1,501 @@
+    }
+
+    /// Return the expression represented by this partial's constraints.
+    pub fn into_expression(self) -> Term {
+        Term::new_temporary(Value::Expression(Operation {
+            operator: Operator::And,
+            args: self
+                .operations
+                .into_iter()
+                .map(|op| Term::new_temporary(Value::Expression(op)))
+                .collect(),
+        }))
+    }
+
+    pub fn clone_with_name(&self, name: Symbol) -> Self {
+        let mut new = self.clone();
+        new.variable = name;
+        new
+    }
+
+    pub fn clone_with_operations(&self, operations: Vec<Operation>) -> Self {
+        let mut new = self.clone();
+        new.operations = operations;
+        new
+    }
+
+    pub fn name(&self) -> &Symbol {
+        &self.variable
+    }
+
+    fn variable_term(&self) -> Term {
+        Term::new_temporary(Value::Variable(sym!("_this")))
+    }
+}
+
+#[derive(Clone)]
+struct IsaConstraintCheck {
+    existing: Vec<Operation>,
+    proposed_tag: Option<Symbol>,
+    result: Option<bool>,
+    last_call_id: u64,
+}
+
+impl IsaConstraintCheck {
+    pub fn new(existing: Vec<Operation>, mut proposed: Operation) -> Self {
+        let right = proposed.args.pop().unwrap();
+        let proposed_tag = if let Value::Pattern(Pattern::Instance(instance)) = right.value() {
+            Some(instance.tag.clone())
+        } else {
+            None
+        };
+
+        Self {
+            existing,
+            proposed_tag,
+            result: None,
+            last_call_id: 0,
+        }
+    }
+
+    /// Check if constraint is compatible with proposed.
+    ///
+    /// Returns: None if compatible, QueryEvent::Done { false } if incompatible,
+    /// or QueryEvent to ask for compatibility.
+    fn check_constraint(
+        &mut self,
+        mut constraint: Operation,
+        counter: &Counter,
+    ) -> Option<QueryEvent> {
+        if constraint.operator != Operator::Isa {
+            return None;
+        }
+
+        let right = constraint.args.pop().unwrap();
+        if let Value::Pattern(Pattern::Instance(instance)) = right.value() {
+            let call_id = counter.next();
+            self.last_call_id = call_id;
+
+            // is_subclass check of instance tag against proposed
+            return Some(QueryEvent::ExternalIsSubclass {
+                call_id,
+                left_class_tag: self.proposed_tag.clone().unwrap(),
+                right_class_tag: instance.tag.clone(),
+            });
+
+            // TODO check fields for compatibility.
+        }
+
+        None
+    }
+}
+
+impl Runnable for IsaConstraintCheck {
+    fn run(&mut self, counter: Counter) -> PolarResult<QueryEvent> {
+        if self.proposed_tag.is_none() {
+            return Ok(QueryEvent::Done { result: true });
+        }
+
+        if let Some(result) = self.result.take() {
+            if !result {
+                return Ok(QueryEvent::Done { result: false });
```

Comment:
What does it mean for the result to be `false` before the `run()` has executed?

Reply:
`self.result` would only be false after an external is subclass check.  This means that the proposed new constraint was not a subclass of an existing constraint.

Reply:
That a child runnable returned `false`, in which case we should, too. But this probably will not happen before `run` has executed; it's re-entered after each event.